### ISSUE TITLE
insertParagraph definition is different from current browsers' behavior when selection is collapsed at end of inline element(s)

### ISSUE THREAD

**Issue by masayuki-nakano** [w3c/editing#172]
Although I'm not sure here is a good place to file an issue of execCommand spec:
https://w3c.github.io/editing/execCommand.html#the-insertparagraph-command

insertParagraph command declares that when selection is started from end of inline element(s), start of range to copy nodes to new container should be next to its next node (16). However, actually, at least Gecko and Blink copies inline elements for keeping same style in new paragraph. For example, |&lt;p>&lt;b>foo[]&lt;/b>&lt;/p>| will be |&lt;p>&lt;p>foo&lt;/b>&lt;/p>&lt;p>&lt;b>[]&lt;br>&lt;/b>&lt;/p>|. I think that this behavior makes sense.

Oddly, if |&lt;b>| is |&lt;a href="">|, Blink doesn't copy it and Gecko didn't do it too (but currently, we changed the behavior accidentally, we will take the old behavior back *1). I feel that this must be better for users since users must not want to keep creating same link in new paragraph in most cases. So, perhaps, this should also be speced.

1: https://bugzilla.mozilla.org/show_bug.cgi?id=1422234

Here are some testcases:

at end of text node:
https://jsfiddle.net/d_toybox/x0kohuo0/8/ (|&lt;b>|)
https://jsfiddle.net/d_toybox/x0kohuo0/9/ (|&lt;a name="foo">|)
https://jsfiddle.net/d_toybox/x0kohuo0/16/ (|&lt;a href="foo">|)

at after text node:
https://jsfiddle.net/d_toybox/x0kohuo0/11/ (|&lt;b>|)
https://jsfiddle.net/d_toybox/x0kohuo0/12/ (|&lt;a href="foo">|)

nested case:
https://jsfiddle.net/d_toybox/x0kohuo0/13/ (|&lt;b>&lt;a href="foo">|)
https://jsfiddle.net/d_toybox/x0kohuo0/14/ (|&lt;b>&lt;i>|)
https://jsfiddle.net/d_toybox/x0kohuo0/15/ (|&lt;a href="foo">&lt;i>|)

**Comment by gked** [comment]
@masayuki-nakano  specing this makes sense. I think, @johanneswilm is the formal spec editor for ExecCommand spec so I would say this GitHub issue tracker as close as it gets to it. :)

I would also propose making a change in the spec to explicitly state assumption for caret position at the time of running insertParagraph command (we are currently assuming that the caret is at the end of string). 

When the caret position is not at the end of the string, expected behavior could be different.

I found few other issues in Chrome using your test cases (I think they are bugs but Blink folks should probably weigh in on this) which have same broken behavior as Gecko.
This is what I tested on Windows 10 recent insider's build for Chrome:
[https://jsfiddle.net/d_toybox/x0kohuo0/9/ (|<a name="foo">|)](https://jsfiddle.net/d_toybox/x0kohuo0/9/ (|<a name="foo">|)) possibly a bug on Chrome
[https://jsfiddle.net/d_toybox/x0kohuo0/10/ (|<a href="foo">|) ](https://jsfiddle.net/d_toybox/x0kohuo0/10/ (|<a href="foo">|) ) possibly a bug on Chrome
nested case:
[https://jsfiddle.net/d_toybox/x0kohuo0/13/ (|<b><a href="foo">|)](https://jsfiddle.net/d_toybox/x0kohuo0/13/ (|<b><a href="foo">|)) possibly a bug on Chrome

**Comment by masayuki-nakano** [comment]
I'll add some tests into editing/data/insertparagraph.js for these cases in a couple of days.

**Comment by johanneswilm** [comment]
@masayuki-nakano Right, and this difference in how `<b>` and `<a>` is something all JS editors deal with in some way or other. Either they hardcode the difference or they create different categories of inline elements -- that that continue if you enter anything at the end of them and those that don't. But this is really something best dealt with in JavaScript because there can be (and are) inline formatting things that the browser doesn't really understand as easily as `<a>`. For example, what should the browser do if the user hits enter in this case:

    <p>Hello <span class="actor-name">Peter[]</span></p>

Additionally, paragraph splitting is best left to JavaScript, because there are 1000s of ways that paragraph splitting and merging could be done, and without domain knowledge of the text's contents, you're really making a blind guess as to what is right.

**Comment by masayuki-nakano** [comment]
> because there can be (and are) inline formatting things that the browser doesn't really understand as easily as `<a>`. For example, what should the browser do if the user hits enter in this case:
> ```
>  <p>Hello <span class="actor-name">Peter[]</span></p>

Well, I don't think that this case should be cared by browsers since contenteditable/designMode do not have built-in UI to modify semantics of each element and such case really depends on what web apps intent. So, if web apps need to provide such additional meaning with `<span>` or something, they should manage by themselves. (So, I believe that contenteditable/designMode are just rich text editor rather than HTML editor.)

> Additionally, paragraph splitting is best left to JavaScript, because there are 1000s of ways that paragraph splitting and merging could be done, and without domain knowledge of the text's contents, you're really making a blind guess as to what is right.

Yeah, but Mozilla needs to keep compatibility with Blink and WebKit as far as possible due to market share difference. So, if we (Mozilla) want to add wpt for such undocumented behavior, we'd like to know what other browser vendors think.  If at least another vendor agree with proposed behavior and that becomes common behavior of some browsers, I believe that that can be included into wpt.

**Comment by johanneswilm** [comment]
> So, I believe that contenteditable/designMode are just rich text editor rather than HTML editor.

Ok, but is there any editor in the world that does not use at least one custom element? I don't think I've seen one that only does bold/italic/links and paragraph breaks. Even if such an editor exists, it will be highly annoying if one cannot extend it to do more things easily but has to implement all logic again. 

I think it's fine that browsers don't want to deal with those extra elements. That's part of the reason we are moving away from execCommand and toward beforeinput events.

> Yeah, but Mozilla needs to keep compatibility with Blink and WebKit as far as possible due to market share difference.

Ok, granted. Anything that makes Firefox and the other browsers behave more similar will generally make it easier to build an editor. So as long as you are just trying to recreate what Chrome does already (and that is reasonable behavior) that sounds fine. But I'm a little afraid we can get distracted by trying to standardize paragraph merging and splitting (which quickly can turn into a very long debate) as well as standardizing execCommands commands that are no longer used. Let's at least make sure we don't forget about the beforeinput event as that will be what really will give the tools needed to build standardized editors across browsers if done right.

**Comment by masayuki-nakano** [comment]
> Ok, but is there any editor in the world that does not use at least one custom element?

Personally, I don't worry about such web apps since such web apps need to manage such custom elements by themselves anyway. Most compatibility issues reported to us are, web developers only tested with Chrome (and perhaps Safari) even though they wrote a code depending on non-standardized behavior. I.e., such web apps are handling input or using existing editing API roughly.

> But I'm a little afraid we can get distracted by trying to standardize paragraph merging and splitting (which quickly can turn into a very long debate) as well as standardizing execCommands commands that are no longer used.

Even if web apps don't use insertParagraph command directly, if they don't override Enter key press, the command is executed indirectly. So, execCommand is used forever indirectly.

**Comment by johanneswilm** [comment]
> Personally, I don't worry about such web apps since

Right, but my question is: Is contenteditable used anywhere in a production level environment that is not a web app with such complexity? In the editing taskforce, we have not come across such a simple editor, with the exception of simple test cases. It is this realization that contenteditable cannot seriously be used by a "simple editor" and consequently is not being used that way which lead us down the path of specifying the beforeinput event.

This does not mean that we don't care about the 'little (wo)man' who just wants to create a simple webpage for her/himself. Several of the professionally used JS editors that exist today are open source and relatively easy to include on a webpage also by people who only recently began learning about how to create web pages.

What we need to worry about is the interface we give to the web developers creating the editors, and they are the ones then providing the tools needed by those creating web pages.

This also has to do with manpower: the browsers all struggle with putting developers on contenteditable, because this is not specifically important to them. But there a number of smaller companies and organizations behind the various JS editors who do have the developers to create editors and the time to listen to end users' complaints and requirements. That's why it makes sense to let these JS devs create their editors without complicating their lives unnecessarily. Should it emerge in a few years that browsers will put many more resources into this, and should some JavaScript editor have become the defacto standard, then it's possible that it makes sense to use reimplement that editor's API in browser native code and ship it like that.

> Even if web apps don't use insertParagraph command directly, if they don't override Enter key press, the command is executed indirectly. So, execCommand is used forever indirectly.

Right, but that is why enter, backspace, delete, pasting, etc. are among the first things to be overriden and handled in JavaScript. Even without a single bug and everything working exactly the same in each browser, as long as there is at least one custom element, it will break the entire thing.
